sockeye/beam_search.py

```diff
@@ -780,30 +826,53 @@ def forward(self,
+            if target_prefix_factors is None:
+                target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
+                                                                                         model_states,
+                                                                                         vocab_slice_ids)
+            else:
+                target_prefix_factor = target_prefix_factors[:, t-1, :] if self.num_target_factors > 1 and t <= target_prefix_factors.size(1) else None
+                target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
+                                                                                         model_states,
+                                                                                         vocab_slice_ids,
+                                                                                         target_prefix_factor)
 
             # (2) Produces the accumulated cost of target words in each row.
             # There is special treatment for finished rows.
             # Finished rows are inf everywhere except column zero, which holds the accumulated model score
             scores, lengths = self._update_scores(target_dists, finished, scores_accumulated,
                                                   lengths, max_output_lengths, pad_dist, eos_dist)
 
+            if target_prefix is not None and t <= target_prefix.size(1):
+                # Make sure search selects the current prefix token by setting the scores of all
+                # other vocabulary items to infinity.
+                target_prefix_in_one_hot = utils.one_hot_encoding_from_prefix(target_prefix[:, t-1:t], self.output_vocab_size)
+                target_prefix_in_one_hot = target_prefix_in_one_hot.expand(-1, self.beam_size, -1).reshape(-1, target_prefix_in_one_hot.size(-1))
+                if vocab_slice_ids is not None:
+                    target_prefix_in_one_hot = pt.index_select(target_prefix_in_one_hot, -1, vocab_slice_ids)
+                scores.masked_fill_(target_prefix_in_one_hot == 0, onp.inf)
+
             # (3) Get beam_size winning hypotheses for each sentence block separately. Only look as
             # far as the active beam size for each sentence.
             if self._sample is not None:
                 best_hyp_indices, best_word_indices, scores_accumulated = self._sample(scores, target_dists, finished)
             else:
                 # On the first timestep, all hypotheses have identical histories, so force topk() to choose extensions
                 # of the first row only by setting all other rows to inf
-                if t == 1:
-                    scores += first_step_mask
+                if target_prefix is None:
+                    scores = scores + first_step_mask if t == 1 else scores
+                else:
```

Comment:
I think it would be nice to have a brief explanation here, something like:
`# while decoding target prefixes, we mask all other hypotheses than the first`

Reply:
Done. Thanks!